langwatch/src/app/api/scenario-events/[[...route]]/scenario-event.repository.ts

```diff
@@ -348,67 +349,205 @@ export class ScenarioEventRepository {
-   * @returns Array of batch run IDs, ordered by most recent first
+   * @param limit - Maximum number of batch run IDs to return
+   * @param cursor - Cursor for pagination (search_after value)
+   * @returns Object containing batch run IDs and pagination info
    * @throws {z.ZodError} If validation fails for projectId or scenarioSetId
    */
   async getBatchRunIdsForScenarioSet({
     projectId,
     scenarioSetId,
+    limit,
+    cursor,
   }: {
     projectId: string;
     scenarioSetId: string;
-  }): Promise<string[]> {
+    limit: number;
+    cursor?: string;
+  }): Promise<{
+    batchRunIds: string[];
+    nextCursor?: string;
+    hasMore: boolean;
+  }> {
     const validatedProjectId = projectIdSchema.parse(projectId);
     const validatedScenarioSetId = scenarioIdSchema.parse(scenarioSetId);
     const client = await this.getClient();
 
+    // Validate and clamp the limit to prevent ES result window issues
+    const actualLimit = Math.min(limit, 20);
+
+    // Parse cursor to get search_after values
+    let searchAfter: any[] | undefined;
+    if (cursor) {
+      try {
+        // Try to decode base64 cursor first (new stable format)
+        const decodedCursor = Buffer.from(cursor, "base64").toString("utf-8");
+        const cursorData = JSON.parse(decodedCursor);
+
+        // Validate cursor shape - expected: [timestamp:number, batchRunId:string, _id?:string]
+        if (
+          Array.isArray(cursorData) &&
+          (cursorData.length === 2 || cursorData.length === 3) &&
+          typeof cursorData[0] === "number" &&
+          typeof cursorData[1] === "string"
+        ) {
+          searchAfter = cursorData;
+        }
+      } catch (e) {
+        // Try legacy JSON cursor format for backward compatibility
+        try {
+          const cursorData = JSON.parse(cursor);
+          if (cursorData.searchAfter && Array.isArray(cursorData.searchAfter)) {
+            searchAfter = cursorData.searchAfter;
+          }
+        } catch (legacyError) {
+          Sentry.captureException(legacyError);
+          // Invalid cursor in both formats, start from beginning
+        }
+      }
+    }
+
+    // Request more items to account for potential deduplication
+    // We need to ensure we get enough unique batch runs after deduplication
+    const requestSize = Math.max(actualLimit * 3, 50); // Request 3x the limit or at least 50
+
     const response = await client.search({
       index: this.indexName,
       body: {
         query: {
           bool: {
-            must: [
+            filter: [
+              { term: { [ES_FIELDS.projectId]: validatedProjectId } },
+              { term: { [ES_FIELDS.scenarioSetId]: validatedScenarioSetId } },
+              { exists: { field: ES_FIELDS.batchRunId } },
+            ],
+          },
+        },
+        sort: [
+          { timestamp: { order: "desc" } },
+          { [ES_FIELDS.batchRunId]: { order: "asc" } }, // Tiebreaker for stable sorting
+        ],
```

Comment:
this should dedupe and ensure we only get the latest doc for each

```suggestion
        ],
        collapse: {
            field: ES_FIELDS.batchRunId,
        },
```